File: backend/app/schemas/alert.py
"""Pydantic alert schemas."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import (
    AlertSource,
    AlertStatus,
    Category,
    Certainty,
    Severity,
    Urgency,
)


class CanonicalAlert(BaseModel):
    """Normalized alert before persistence."""

    model_config = ConfigDict(from_attributes=True)

    source: AlertSource
    source_alert_id: str
    source_url: str | None = None
    title: str
    description: str | None = None
    instruction: str | None = None
    country_code: str | None = None
    country_name: str | None = None
    region: str | None = None
    location_name: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    geometry: dict[str, Any] | None = None
    category: Category
    event_type: str | None = None
    severity: Severity
    urgency: Urgency | None = None
    certainty: Certainty | None = None
    status: AlertStatus = AlertStatus.ACTUAL
    language: str | None = None
    issued_at: datetime
    effective_at: datetime | None = None
    starts_at: datetime | None = None
    expires_at: datetime | None = None
    updated_at_source: datetime | None = None
    raw_payload: dict[str, Any]


class AlertResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source: AlertSource
    source_alert_id: str
    source_url: str | None = None
    title: str
    description: str | None = None
    instruction: str | None = None
    country_code: str | None = None
    country_name: str | None = None
    region: str | None = None
    location_name: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    geometry: dict[str, Any] | None = None
    category: Category
    event_type: str | None = None
    severity: Severity
    urgency: Urgency | None = None
    certainty: Certainty | None = None
    status: AlertStatus
    language: str | None = None
    issued_at: datetime
    effective_at: datetime | None = None
    starts_at: datetime | None = None
    expires_at: datetime | None = None
    updated_at_source: datetime | None = None
    ingested_at: datetime
    last_seen_at: datetime
    raw_payload: dict[str, Any] | None = None
    fingerprint: str
    is_active: bool
    ingest_mode: str = "live"


class AlertListResponse(BaseModel):
    items: list[AlertResponse]
    total: int
    limit: int
    offset: int


class AlertQueryParams(BaseModel):
    source: AlertSource | None = None
    country: str | None = None
    category: Category | None = None
    severity: Severity | None = None
    active: bool = True
    issued_after: datetime | None = None
    issued_before: datetime | None = None
    bounding_box: str | None = None
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)
    include_raw: bool = False

File: frontend/types/alert.ts
export type AlertSource = "nina" | "gdacs" | "noaa";

/** Observed-event sources (Phase 1–2); not used for alert filters yet. */
export type ObservedEventSource = "usgs" | "eonet" | "noaa_swpc";

export type DataSource = AlertSource | ObservedEventSource;

export type SpatialScope =
  | "local"
  | "regional"
  | "continental"
  | "global"
  | "orbital";

export type Severity = "unknown" | "minor" | "moderate" | "severe" | "extreme";

export type Urgency =
  | "immediate"
  | "expected"
  | "future"
  | "past"
  | "unknown";

export type Certainty =
  | "observed"
  | "likely"
  | "possible"
  | "unlikely"
  | "unknown";

export type Category =
  | "weather"
  | "flood"
  | "wildfire"
  | "earthquake"
  | "volcano"
  | "tsunami"
  | "health"
  | "civil"
  | "infrastructure"
  | "environmental"
  | "other";

export type AlertStatus = "actual" | "exercise" | "test" | "draft" | "unknown";

export interface Alert {
  id: string;
  source: AlertSource;
  source_alert_id: string;
  source_url: string | null;
  title: string;
  description: string | null;
  instruction: string | null;
  country_code: string | null;
  country_name: string | null;
  region: string | null;
  location_name: string | null;
  latitude: number | null;
  longitude: number | null;
  geometry: GeoJSON.Geometry | null;
  category: Category;
  event_type: string | null;
  severity: Severity;
  urgency: Urgency | null;
  certainty: Certainty | null;
  status: AlertStatus;
  language: string | null;
  issued_at: string;
  effective_at: string | null;
  starts_at: string | null;
  expires_at: string | null;
  updated_at_source: string | null;
  ingested_at: string;
  last_seen_at: string;
  raw_payload?: Record<string, unknown> | null;
  fingerprint: string;
  is_active: boolean;
  ingest_mode?: "live" | "fixture" | "showcase";
}

export interface ObservedEvent {
  id: string;
  source: ObservedEventSource;
  source_event_id: string;
  source_url: string | null;
  title: string;
  description: string | null;
  event_type: string | null;
  category: Category;
  severity: Severity;
  status: string;
  confidence: string;
  location_name: string | null;
  latitude: number | null;
  longitude: number | null;
  geometry: GeoJSON.Geometry | null;
  spatial_scope: SpatialScope;
  affected_latitude_min: number | null;
  affected_latitude_max: number | null;
  issued_at: string;
  starts_at: string | null;
  ends_at: string | null;
  source_metadata?: Record<string, unknown> | null;
  fingerprint: string;
  is_active: boolean;
  ingest_mode?: "live" | "fixture" | "showcase";
}

export interface AlertListResponse {
  items: Alert[];
  total: number;
  limit: number;
  offset: number;
}

export interface AlertFilters {
  source?: AlertSource;
  country?: string;
  category?: Category;
  severity?: Severity;
  active?: boolean;
  issued_after?: string;
  issued_before?: string;
  bounding_box?: string;
  limit?: number;
  offset?: number;
}

export interface ObservedEventFilters {
  source?: ObservedEventSource;
  category?: Category;
  severity?: Severity;
  active?: boolean;
  bounding_box?: string;
  limit?: number;
  offset?: number;
}
